File: openless-all/app/src/lib/ipc/channels.ts
// 渠道配置的 typed IPC：排序、启停、名称和手动验证结果。
// Core 选择排序后的第一个启用渠道；凭据通过 readCredential/setCredential 按渠道 ID 读写。

import { invokeOrMock } from './shared';
import { mockCredentialValues } from './mock-data';

export type ChannelKind = 'llm' | 'asr';

export interface ChannelTestResult {
  ok: boolean;
  latencyMs: number | null;
  /** Unix 秒（后端时钟，前端不要自己生成）。 */
  at: number;
  error: string | null;
}

export interface Channel {
  id: string;
  /** 用户取的名字；空串表示未命名，由 UI 回落到 preset 显示名。 */
  name: string;
  /** 厂商 id —— 决定协议与表单形状，与 id 相互独立。 */
  providerType: string;
  enabled: boolean;
  order: number;
  lastTest: ChannelTestResult | null;
}

// 浏览器（非 Tauri）下的样例数据，让 `npm run dev` 能预览列表的四种状态：
// 生效中 / 备用 / 测试失败标红 / 已关闭沉底。
const mockChannels: Record<ChannelKind, Channel[]> = {
  llm: [
    {
      id: 'siliconflow',
      name: '硅基流动-主号',
      providerType: 'siliconflow',
      enabled: true,
      order: 0,
      lastTest: { ok: true, latencyMs: 284, at: Math.floor(Date.now() / 1000) - 90, error: null },
    },
    {
      id: 'ark',
      name: '',
      providerType: 'ark',
      enabled: true,
      order: 1,
      lastTest: null,
    },
    {
      id: 'openai',
      name: 'OpenAI-备用',
      providerType: 'openai',
      enabled: false,
      order: 2,
      lastTest: {
        ok: false,
        latencyMs: null,
        at: Math.floor(Date.now() / 1000) - 3600,
        error: 'providerHttpStatus:401',
      },
    },
  ],
  asr: [
    {
      id: 'volcengine',
      name: '',
      providerType: 'volcengine',
      enabled: true,
      order: 0,
      lastTest: { ok: true, latencyMs: 143, at: Math.floor(Date.now() / 1000) - 20, error: null },
    },
    {
      id: 'groq',
      name: 'Groq-白嫖号',
      providerType: 'groq',
      enabled: true,
      order: 1,
      lastTest: null,
    },
    {
      id: 'orcarouter-asr',
      name: 'OrcaRouter-ASR',
      providerType: 'orcarouter',
      enabled: false,
      order: 2,
      lastTest: null,
    },
  ],
};

export function listChannels(kind: ChannelKind): Promise<Channel[]> {
  return invokeOrMock('list_channels', { kind }, () => mockChannels[kind]);
}

export function invalidateMockChannelTest(id: string): void {
  const channel = mockChannels.llm.find((channel) => channel.id === id);
  if (channel) channel.lastTest = null;
}

export function invalidateMockChannelTests(kind: ChannelKind): void {
  for (const channel of mockChannels[kind]) channel.lastTest = null;
}

/** 返回后端分配的渠道 id。 */
export function createChannel(
  kind: ChannelKind,
  providerType: string,
  name: string,
): Promise<string> {
  return invokeOrMock('create_channel', { kind, providerType, name }, () => {
    const id = `${providerType}-${Date.now()}-${mockChannels[kind].length}`;
    mockChannels[kind].push({
      id,
      name,
      providerType,
      enabled: true,
      order: mockChannels[kind].length,
      lastTest: null,
    });
    return id;
  });
}

/** 在已建好的草稿卡片上换供应商（单弹窗添加流程的常规操作）。 */
export function setChannelProviderType(
  kind: ChannelKind,
  id: string,
  providerType: string,
): Promise<void> {
  return invokeOrMock('set_channel_provider_type', { kind, id, providerType }, () => {
    const channel = mockChannels[kind].find((channel) => channel.id === id);
    if (channel && channel.providerType !== providerType) {
      channel.providerType = providerType;
      channel.lastTest = null;
      if (kind === 'llm') mockCredentialValues.delete(`${id}:ark.request_format`);
    }
  });
}

/** 关闭添加弹窗时回收没填任何东西的草稿；返回是否真的删了。 */
export function deleteChannelIfBlank(kind: ChannelKind, id: string): Promise<boolean> {
  return invokeOrMock('delete_channel_if_blank', { kind, id }, () => {
    const channel = mockChannels[kind].find((channel) => channel.id === id);
    const prefix = `${id}:`;
    const hasCredentials = [...mockCredentialValues].some(
      ([key, value]) => key.startsWith(prefix) && value.length > 0,
    );
    if (!channel || channel.name.trim() || hasCredentials) return false;
    mockChannels[kind] = mockChannels[kind].filter((channel) => channel.id !== id);
    for (const key of mockCredentialValues.keys()) {
      if (key.startsWith(prefix)) mockCredentialValues.delete(key);
    }
    return true;
  });
}

export function renameChannel(kind: ChannelKind, id: string, name: string): Promise<void> {
  return invokeOrMock('rename_channel', { kind, id, name }, () => {
    const channel = mockChannels[kind].find((channel) => channel.id === id);
    if (channel) channel.name = name;
    return undefined;
  });
}

export function deleteChannel(kind: ChannelKind, id: string): Promise<void> {
  return invokeOrMock('delete_channel', { kind, id }, () => {
    mockChannels[kind] = mockChannels[kind].filter((channel) => channel.id !== id);
    for (const key of mockCredentialValues.keys())
      if (key.startsWith(`${id}:`)) mockCredentialValues.delete(key);
  });
}

export function setChannelEnabled(kind: ChannelKind, id: string, enabled: boolean): Promise<void> {
  return invokeOrMock('set_channel_enabled', { kind, id, enabled }, () => {
    const channel = mockChannels[kind].find((channel) => channel.id === id);
    if (channel) channel.enabled = enabled;
    return undefined;
  });
}

/** ids 是拖拽后的完整顺序；后端会把未提及的渠道排到末尾。 */
export function reorderChannels(kind: ChannelKind, ids: string[]): Promise<void> {
  return invokeOrMock('reorder_channels', { kind, ids }, () => {
    // mock 也要真的重排：否则浏览器预览里松手后顺序被 listChannels 拉回原样，
    // 看着就像"拖拽坏了"，而真机是好的。
    const list = mockChannels[kind];
    const ordered = ids
      .map((id) => list.find((c) => c.id === id))
      .filter((c): c is Channel => Boolean(c));
    const rest = list.filter((c) => !ids.includes(c.id));
    mockChannels[kind] = [...ordered, ...rest].map((c, index) => ({
      ...c,
      order: index,
    }));
    return undefined;
  });
}

export function recordChannelTest(
  kind: ChannelKind,
  id: string,
  ok: boolean,
  latencyMs: number | null,
  error: string | null,
): Promise<void> {
  return invokeOrMock('record_channel_test', { kind, id, ok, latencyMs, error }, () => {
    const channel = mockChannels[kind].find((channel) => channel.id === id);
    if (channel) channel.lastTest = { ok, latencyMs, error, at: Math.floor(Date.now() / 1000) };
  });
}

File: openless-all/app/src-tauri/src/commands/channels.rs
//! 渠道卡片管理的 IPC 面。
//!
//! 一张卡片 = 一份可命名、可排序、可开关的供应商配置。同一家厂商可以有多张卡片
//! （多把 key），此时渠道 id 与 `providerType` 分离 —— 前者是 map key，后者决定
//! 协议路由。详见 `persistence::credentials` 里 `ChannelMeta` 的说明。
//!
//! 凭据本身不走这里：前端按渠道 id 调 `read_credential` / `set_credential`
//! （`provider` 参数传渠道 id），避免密钥随列表批量出栈。

use super::*;
use openless_core::{ChannelKind, ChannelSummary};

fn parse_kind(kind: &str) -> Result<ChannelKind, String> {
    ChannelKind::parse(kind).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_channels(
    core: CoreState<'_>,
    window: Window,
    kind: String,
) -> Result<Vec<ChannelSummary>, String> {
    ensure_main_window(&window)?;
    core.list_channels(parse_kind(&kind)?)
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn create_channel(
    core: CoreState<'_>,
    window: Window,
    kind: String,
    provider_type: String,
    name: String,
) -> Result<String, String> {
    ensure_main_window(&window)?;
    core.create_channel(parse_kind(&kind)?, provider_type, name)
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn set_channel_provider_type(
    core: CoreState<'_>,
    window: Window,
    kind: String,
    id: String,
    provider_type: String,
) -> Result<(), String> {
    ensure_main_window(&window)?;
    core.set_channel_provider_type(parse_kind(&kind)?, id, provider_type)
        .await
        .map_err(|error| error.to_string())
}

/// 关闭「添加渠道」弹窗时回收没填任何内容的草稿卡片；返回是否真的删了。
#[tauri::command]
pub async fn delete_channel_if_blank(
    core: CoreState<'_>,
    window: Window,
    kind: String,
    id: String,
) -> Result<bool, String> {
    ensure_main_window(&window)?;
    core.delete_channel_if_blank(parse_kind(&kind)?, id)
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn rename_channel(
    core: CoreState<'_>,
    window: Window,
    kind: String,
    id: String,
    name: String,
) -> Result<(), String> {
    ensure_main_window(&window)?;
    core.rename_channel(parse_kind(&kind)?, id, name)
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn delete_channel(
    core: CoreState<'_>,
    window: Window,
    kind: String,
    id: String,
) -> Result<(), String> {
    ensure_main_window(&window)?;
    core.delete_channel(parse_kind(&kind)?, id)
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn set_channel_enabled(
    core: CoreState<'_>,
    window: Window,
    kind: String,
    id: String,
    enabled: bool,
) -> Result<(), String> {
    ensure_main_window(&window)?;
    core.set_channel_enabled(parse_kind(&kind)?, id, enabled)
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn reorder_channels(
    core: CoreState<'_>,
    window: Window,
    kind: String,
    ids: Vec<String>,
) -> Result<(), String> {
    ensure_main_window(&window)?;
    core.reorder_channels(parse_kind(&kind)?, ids)
        .await
        .map_err(|error| error.to_string())
}

/// 记录一次「测试连通」的结果，供卡片显示延迟或标红。
///
/// 时间戳在后端取，不信任前端传入 —— 前端时钟错乱会让"3 分钟前"显示成负数。
#[tauri::command]
pub async fn record_channel_test(
    core: CoreState<'_>,
    window: Window,
    kind: String,
    id: String,
    ok: bool,
    latency_ms: Option<u32>,
    error: Option<String>,
) -> Result<(), String> {
    ensure_main_window(&window)?;
    core.record_channel_test(parse_kind(&kind)?, id, ok, latency_ms, error)
        .await
        .map_err(|error| error.to_string())
}
